### ISSUE TITLE
About multi-speaker datasets and tacotron2...

### ISSUE THREAD

**Issue by samuel-lunii** [TensorSpeech/TensorFlowTTS#644]
Hi ! 
I would like to ask a few questions about multi-speaker datasets. [This thread](https://github.com/TensorSpeech/TensorFlowTTS/issues/296#issuecomment-741850534) gives good insights about what is needed for transfer learning with short duration for each speaker in a multi-speaker dataset using FastSpeech.

My questions are about Tacotron2.

- What is the minimal total duration needed to train a good multi-speaker model from scratch ?
- What is the minimal duration needed for each speaker in the dataset ?
- Would it be good practice to use large durations (e.g. 1 or 2 speakers with duration > 20 hours) in combination with shorter ones (e.g. several speakers with duration ~ 1 hour) to train a single multi-speaker model ? Or is it better to perform transfer learning with short durations from a model trained on large durations ? 

Thanks !

**Comment by dathudeptrai** [comment]
@ZDisket do you have any ideas?

**Comment by ZDisket** [comment]
@dathudeptrai @samuel-lunii I haven't had much success training multispeaker Tacotron2 as the attention mechanism becomes unable to learn alignment when it is multispeaker, even for datasets that are big, good and train successfully on FastSpeech2, so I would suggest using that for multispeaker instead. Maybe we can fix the issues of fixed durations for outputs by using the stochastic duration predictor from VITS, which has variations even with the same input, instead of the regular one in FastSpeech2. I'll try to answer every question

> What is the minimal total duration needed to train a good multi-speaker model from scratch ?

I would say about 15 to 20 hours total

> What is the minimal duration needed for each speaker in the dataset ?

In my limited experiments, those that perform well have at least 30 minutes

>Would it be good practice to use large durations (e.g. 1 or 2 speakers with duration > 20 hours) in combination with shorter ones (e.g. several speakers with duration ~ 1 hour) to train a single multi-speaker model ? Or is it better to perform transfer learning with short durations from a model trained on large durations ?

You should include the speakers with little data into the big model, as when you try to fine-tune, unless it's the same amount of speakers, the embedding layer will be dropped due to incompatibility, and that will mess the new model up.

**Comment by samuel-lunii** [comment]
@ZDisket 
Thanks for your answers.

> I haven't had much success training multispeaker Tacotron2 as the attention mechanism becomes unable to learn alignment when it is multispeaker, even for datasets that are big, good and train successfully on FastSpeech2, so I would suggest using that for multispeaker instead.

So I guess your answers are about FastSpeech2 ? I actually want to use Tacotron2 for expressive speech synthesis.

As mentioned in #628, I think it could be worth adapting the Tacotron2 model in a way similar to the [prosody paper](http://proceedings.mlr.press/v80/skerry-ryan18a/skerry-ryan18a.pdf) (also used in the [GST paper](http://proceedings.mlr.press/v80/wang18h/wang18h.pdf)), where the speaker embedding is broadcasted to match the input text sequence length and then concatenated to the encoder output. I did have correct results by doing so with [keithito's repo](https://github.com/keithito/tacotron). I will let you know if I meet any success :) 

By the way, @dathudeptrai @ZDisket  
I see that the speaker embeddings are used [before](https://github.com/TensorSpeech/TensorFlowTTS/blob/29595017ab1b2a3bdef444b2195648383fa5ed05/tensorflow_tts/models/tacotron2.py#L172-L178) and [after](https://github.com/TensorSpeech/TensorFlowTTS/blob/29595017ab1b2a3bdef444b2195648383fa5ed05/tensorflow_tts/models/tacotron2.py#L254-L264) the encoder. Is there a particular reason to do so ? Have you tried to use either the one or the other independently ?

**Comment by dathudeptrai** [comment]
@samuel-lunii 

> I see that the speaker embeddings are used before and after the encoder. Is there a particular reason to do so ? Have you tried to use either the one or the other independently ?

Add before ->>> speaker-dependently encoder
Add after ->>> speaker-dependently decoder 

You can imagine it as Resnet.

**Comment by ZDisket** [comment]
@samuel-lunii 

> I actually want to use Tacotron2 for expressive speech synthesis.

Define "expressive".  FastSpeech2 is good at expressiveness, it's just the deterministic duration predictions that are the problem. I've had success adding an emotion embedding (a copy of the speaker embedding) for training with a multispeaker multi-emotion dataset that is categorized by those.

**Comment by samuel-lunii** [comment]
@ZDisket 
Ok good to know, thanks :) I will definitely have a go with multi-speaker FasteSpeech2.
However, I would like to use "GST like" architecture, in order to avoid emotion labelling...

**Comment by ZDisket** [comment]
@samuel-lunii
>to avoid emotion labelling...

 Someone I know used an automatic sentiment detection from text labeling model and used its output as emotion embedding.

**Comment by samuel-lunii** [comment]
@ZDisket 
Yes, I was planning on doing something similar too :) I will follow your advice and try all this with FS2.
